### ISSUE TITLE
Wordpress cached data unserialized as string instead of array after switching redis-cache-pro from standard serializer to igbinary

### ISSUE THREAD

**Issue by dmitriyelagin** [igbinary/igbinary#376]
lighttpd 1.4.69, PHP-fpm 8.2.3, debian 11, 10.5.18-MariaDB, redis-cache-pro - compression lz4
All packages are installed via apt

php info
igbinary support - enabled igbinary version 3.2.13
redis - Available serializers php, json, igbinary
session - Registered serializer handlers php_serialize php php_binary igbinary msgpack
session.serialize_handler igbinary igbinary

php -i | grep igbinary
/etc/php/8.2/cli/conf.d/20-igbinary.ini,
igbinary
igbinary support => enabled
igbinary version => 3.2.13
igbinary APCu serializer ABI => 0
igbinary session support => yes
igbinary.compact_strings => On => On
igbinary support => yes
Available serializers => php, json, igbinary
Registered serializer handlers => php_serialize php php_binary igbinary msgpack
session.serialize_handler => igbinary => igbinary

If you use php as the default serializer in the configuration, then everything works very quickly. But I read that igbinary_serialize works several times faster?

`define( 'WP_REDIS_CONFIG', [
'token' => '',
'host' => '127.0.0.1',
'port' => 6379,
'database' => 5,
'maxttl' => 3600 * 24 * 7,
'timeout' => 0.5,
'read_timeout' => 0.5,
'retry_interval' => 10,
'retries' => 3,
//'serializer' => 'igbinary',
'compression' => 'lz4',
'async_flush' => true,
'client' => 'phpredis',
'split_alloptions' => true,
'shared' => true,
'prefetch' => true,
'debug' => false,
'save_commands' => false,
]);
define('WP_REDIS_DISABLED', getenv('WP_REDIS_DISABLED') ?: false);`

But when I turn on 'serializer' = > 'igbinary', in the config, I get a critical error in Wordpress.

`2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:PHP message: PHP Fatal error:  Uncaught TypeError: array_keys(): Argument #1 ($array) must be of type array, string given in /www/free/wp-includes/class-wp-roles.php:291
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:Stack trace:
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:#0 /www/free/wp-includes/class-wp-roles.php(291): array_keys()
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:#1 /www/free/wp-includes/class-wp-roles.php(332): WP_Roles->init_roles()
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:#2 /www/free/wp-includes/class-wp-roles.php(91): WP_Roles->for_site()
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:#3 /www/free/wp-settings.php(542): WP_Roles->__construct()
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:#4 /www/free/wp-config.php(101): require_once('...')
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:#5 /www/free/wp-load.php(50): require_once('...')
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:#6 /www/free/wp-admin/admin-post.php(19): require_once('...')
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:#7 {main}
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:  thrown in /www/free/wp-includes/class-wp-roles.php on line 291; PHP message: PHP Fatal error:  Uncaught TypeError: array_keys(): Argument #1 ($array) must be of type array, string given in /www/free/wp-includes/class-wp-roles.php:291
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:Stack trace:
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:#0 /www/free/wp-includes/class-wp-roles.php(291): array_keys()
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:#1 /www/free/wp-includes/class-wp-roles.php(332): WP_Roles->init_roles()
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:#2 /www/free/wp-includes/class-wp-roles.php(91): WP_Roles->for_site()
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:#3 /www/free/wp-includes/capabilities.php(997): WP_Roles->__construct()
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:#4 /www/free/wp-includes/class-wp-user.php(510): wp_roles()
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:#5 /www/free/wp-includes/class-wp-user.php(877): WP_User->get_role_caps()
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:#6 /www/free/wp-includes/class-wp-user.php(178): WP_User->for_site()
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:#7 /www/free/wp-includes/pluggable.php(109): WP_User->init()
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:#8 /www/free/wp-includes/pluggable.php(737): get_user_by()
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:#9 /www/free/wp-includes/class-wp-hook.php(308): wp_validate_auth_cookie()
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:#10 /www/free/wp-includes/plugin.php(205): WP_Hook->apply_filters()
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:#11 /www/free/wp-includes/user.php(3604): apply_filters()
2023-03-08 21:09:47: (mod_fastcgi.c.449) FastCGI-stderr:#12 /www/free/wp-includes/p...
2023-03-08 21:09:49: (mod_fastcgi.c.449) FastCGI-stderr:PHP message: PHP Fatal error:  Uncaught TypeError: array_keys(): Argument #1 ($array) must be of type array, string given in /www/free/wp-includes/class-wp-roles.php:291`

**Comment by TysonAndre** [comment]
Can you create a minimal, self-contained repo or example file to reproduce this issue? (If possible, the an object/array which when serialized with igbinary, results in an array becoming a non-array)

I'm not familiar with where the redis cache is being used in https://github.com/WordPress/WordPress/blob/master/wp-includes/class-wp-roles.php

I'm also not familiar with https://en-ca.wordpress.org/plugins/redis-cache/ , and a standalone example would be easier to work with

Also, what is the string it's being called with? Can you include it or steps to create a similar string that reproduces the issue?

**Comment by TysonAndre** [comment]
https://github.com/rhubarbgroup/redis-cache/blob/83c771dfc5bedc807133330080598553deb65dd9/includes/object-cache.php#L2691-L2698

Oh. Actually, it turned out to be obvious. It's a bug in redis-cache-pro. redis-cache-pro  is **unconditionally** trying to unserialize the data it retrieved from redis (PHP serialize()) with igbinary (igbinary unserialize()). That fails, so it returns the original string, probably of the form `a:...{...}`

It wouldn't happen in installations that started with an empty redis cache or cleared it (not sure if that would cause issues for you)

It should be doing something similar to phpredis instead, and checking if `substr( $original, 0, 4) === "\0\0\0\0"` or something similar, but it isn't.


```c
        case REDIS_SERIALIZER_IGBINARY:
#ifdef HAVE_REDIS_IGBINARY
            /*
             * Check if the given string starts with an igbinary header.
             *
             * A modern igbinary string consists of the following format:
             *
             * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-
             * | header (4) | type (1) | ... (n) |  NUL (1) |
             * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-
             *
             * With header being either 0x00000001 or 0x00000002
             * (encoded as big endian).
             *
             * Not all versions contain the trailing NULL byte though, so
             * do not check for that.
             */
            if (val_len < 5
                    || (memcmp(val, "\x00\x00\x00\x01", 4) != 0
                    && memcmp(val, "\x00\x00\x00\x02", 4) != 0))
            {
                /* This is most definitely not an igbinary string, so do
                   not try to unserialize this as one. */
                break;
            }

            ret = !igbinary_unserialize((const uint8_t *)val, (size_t)val_len, z_ret);
#endif
            break;
```

**Comment by TysonAndre** [comment]
Never mind, it isn't that, it'd return false if igbinary is loaded.

What are the values of `WP_REDIS_*` constants you have set up or the cache configuration?

What were your original settings (e.g. did compression change?)

**Comment by TysonAndre** [comment]
Which of the advanced options are you using? https://github.com/rhubarbgroup/redis-cache/blob/77fcd714a58806819d85c4a017a8588e98f5fec2/README.md - WP_REDIS_SERIALIZER or WP_REDIS_IGBINARY? (I assume it's WP_REDIS_SERIALIZER and somehow extracted from WP_REDIS_CONFIG)

Does the issue happen with an empty cache?

**Comment by TysonAndre** [comment]
Does the issue happen with an empty cache?

Are you running more than one server? If so, did you upgrade all of the servers? Does the issue happen all of the time?

What redis-cache-pro version is this?

**Comment by TysonAndre** [comment]
> But when I turn on 'serializer' = > 'igbinary', in the config, I get a critical error in Wordpress.

Did you restart apache?

**Comment by dmitriyelagin** [comment]
Thanks for the information, but it's a little hard for me. Tell me what information I can provide?
I have one VPS. Yes, every time I cleared all caches (object cache, wp-rocket pro cache). I use lighttpd 1.4.69, yes I restarted it, as well as php8.2-fpm and redis
https://objectcache.pro/docs/data-encoding/
v1.18.0 Object Cache Pro
php8.2-redis 5.3.7 debian 11

**Comment by dmitriyelagin** [comment]
> It wouldn't happen in installations that started with an empty redis cache or cleared it (not sure if that would cause issues for you)

Thanks so much for your help!
My actions are based on the example of one site.
I entered the root of the site
cd /www/tube
Changed to wp-config
'serializer' => 'igbinary',
and added a prefex
'prefix' => 'tube',
Save
wp redis flush --allow-root
And he also cleared the wp-rocket pro cache in the site admin panel.
This solved the problem.
serializer igbinary
compression lz4
